### DISCUSSION TITLE
DataTemplate is not applied

### DISCUSSION THREAD

**Discussion by sam-wheat** [AvaloniaUI/Avalonia#8387] [Q&A]
My control - `DimensionButton`  - contains another control I wrote called `DropDownButton`

In the style for `DimensionButton` I define a `DataTemplate` and use it to set the `ItemTemplate` property on `DropDownButton`.  

The problem is that the `DataTemplate` is never used.  I can see that `ItemTemplate` and `DataContext` are set correctly on `DropDownButton` using the Diagnostics tool.

What am I doing wrong?

Avalonia repo:
https://github.com/leaderanalytics/LeaderPivot.XAML.Avalonia

Working WPF version of the same control for comparison:
https://github.com/leaderanalytics/LeaderPivot.XAML.WPF

DimensionButton:

		<Style Selector="local|DimensionButton">
		<Style.Resources>
			<DataTemplate DataType="ListBoxItem" x:Key="DimensionButtonItemTemplate">
				//... This is the DataTemplate that should be used
			</DataTemplate>
		</Style.Resources>

		
		<Setter Property="Template">
			<Setter.Value>
				<ControlTemplate TargetType="{x:Type local:DimensionButton}">
					<local:DropDownButton x:Name="dropDownButton1"
						DataContext="{Binding $parent[local:DimensionButton]}"
						ItemTemplate="{StaticResource DimensionButtonItemTemplate}"	// ItemTemplate set here
						Command="{Binding Command}"
						Content="{Binding Content}"
						SelectedItem="{Binding SelectedItem, Mode=TwoWay}"
						ButtonStyle="{Binding ButtonStyle}"
						ListBoxStyle="{Binding ListBoxStyle}"
						ListBoxItemStyle="{Binding ListBoxItemStyle}"
						PopupStyle="{Binding PopupStyle}">

						<local:DropDownButton.ItemsSource>
							<sc:ArrayList>
								<x:Static Member="local:DimensionAction.SortAscending"/>
								<x:Static Member="local:DimensionAction.SortDescending"/>
								<x:Static Member="local:DimensionAction.Hide"/>
							</sc:ArrayList>
						</local:DropDownButton.ItemsSource>
						<i:Interaction.Behaviors>
							<ia:EventTriggerBehavior EventName="SelectionChanged" SourceObject="{Binding #dropDownButton1}" >
								<ia:InvokeCommandAction Command="{Binding SelectionChangedCommand}" CommandParameter="{Binding CommandParameter}"  />
							</ia:EventTriggerBehavior>
						</i:Interaction.Behaviors>
					</local:DropDownButton>
				</ControlTemplate>
			</Setter.Value>
		</Setter>
		
	</Style>

	
Dropdown button:

	<Style Selector="local|DropDownButton">
		<Setter Property="Template">
			<ControlTemplate TargetType="{x:Type local:DropDownButton}" >
				<StackPanel x:Name="sp1" Orientation="Vertical" DataContext="{Binding RelativeSource={RelativeSource AncestorType=local:DropDownButton}}">
						
					<Button x:Name="MartyMcButton" Classes.myClass="{TemplateBinding ButtonStyle}" Content="{TemplateBinding Content}" Command="{Binding ToggleDropDownCommand}"></Button>
					<Popup IsOpen="{Binding IsDropDownOpen}" PlacementTarget="{Binding ElementName=MartyMcButton}" Classes.myClass="{Binding PopupStyle}" x:Name="btnPopup" >
						<i:Interaction.Behaviors>
							<ia:EventTriggerBehavior EventName="PointerLeave" SourceObject="{Binding #btnPopup}" >
								<ia:InvokeCommandAction Command="{Binding MouseLeaveCommand}" />
							</ia:EventTriggerBehavior>
						</i:Interaction.Behaviors>
						<Border Margin="0,0,8,8" Padding="{Binding PopupPadding}" >
							<ListBox x:Name="PART_ListBox" 
                                        Items="{Binding ItemsSource}"
                                        SelectedItem="{Binding SelectedItem, Mode=TwoWay}"
                                        ItemTemplate="{Binding ItemTemplate}"						// ItemTemplate set here
                                        Classes.myClass="{TemplateBinding ListBoxStyle}" >
								<i:Interaction.Behaviors>
									<ia:EventTriggerBehavior EventName="SelectionChanged" SourceObject="PART_ListBox" >
										<ia:ChangePropertyAction PropertyName="IsDropDownOpen" TargetObject="{Binding $parent[local:DropDownButton]}" Value="False"/>
										<ia:InvokeCommandAction Command="{Binding SelectionChangedCommand}"  />
									</ia:EventTriggerBehavior>
								</i:Interaction.Behaviors>
							</ListBox>
						</Border>
					</Popup>
				</StackPanel>
			</ControlTemplate>
		</Setter>
	</Style>
	
	public class DropDownButton : ContentControl, INotifyPropertyChanged
	{
		//...
		
		public DataTemplate ItemTemplate
		{
			get { return (DataTemplate)GetValue(ItemTemplateProperty); }
			set { SetValue(ItemTemplateProperty, value); }
		}

		public static readonly AvaloniaProperty ItemTemplateProperty =
			AvaloniaProperty.Register<DropDownButton,DataTemplate>(nameof(ItemTemplate));
			
		//...
	}
	
![Capture](https://user-images.githubusercontent.com/16885210/175780103-7fb603c7-0715-4f57-bdfe-b50269d1f8cf.PNG)

**Comment by maxkatz6** [comment]
Your repository in attachment doesn't contain these controls.

**Reply by sam-wheat** [reply]
Sorry I also have another remote and I forget to sync with Github.  Please try again.

**Comment by maxkatz6** [comment]
Your template has `DataType="ListBoxItem"`. But why would you expect ListBoxItem to be a data type there?
Your items are of type `local:DimensionAction` as I can see.

Furthermore, DataTemplate by itself uses TemplatedParent, which causes DimensionEventCheckedConverter to fail with "Could not bind to Dimension" exception. TemplatedParent works with Control.Template property and not ItemTemplate from what I know. `AncestorType=local:DimensionButton` should be fine though, or Avalonia syntax `{Binding $parent[local:DimensionButton].Dimension}`.

**Reply by maxkatz6** [reply]
Checking your code, it seems when converter throws an exception for DirectProperty binding, it's not properly handled by Avalonia. So this "Could not bind to Dimension" error was hidden in stacktrace. Fix https://github.com/AvaloniaUI/Avalonia/pull/8391

**Reply by sam-wheat** [reply]
Very big thank you for putting me on the path to enlightenment.  I'm not out of the woods yet but progress has been made thanks to you.

**Reply by sam-wheat** [reply]
@maxkatz6 I am sorry to bother you with this again but I am stuck.  Commented lines don't build, `RelativeSource` builds but does not work.

It is interesting to note that the binding for CheckBox Command that references `$parent[local:DimensionButton]` builds and runs.
However, this syntax: `<Binding Path ="{$parent[local:DimensionButton].Dimension}" ... />` for Checkbox.IsChecked fails to build with "Unable to resolve type DimensionButton].Dimension from namespace".  This error message is an exact copy - note missing left bracket and missing namespace. 

[Code repo](https://github.com/leaderanalytics/LeaderPivot.XAML.Avalonia)


	<DataTemplate DataType="local:DimensionAction" x:Key="DimensionButtonItemTemplate">
		<Border Margin="2,3,2,3">
			<CheckBox HorizontalAlignment="Left" VerticalAlignment="Center"
						  Command="{Binding $parent[local:DimensionButton], Path=CheckboxCheckedCommand}"
						  CommandParameter="{Binding}"
						  Content="{Binding, Converter={StaticResource DimensionActionDescriptionConverter}}">

				<CheckBox.IsChecked>
					<MultiBinding Converter="{StaticResource DimensionEventCheckedConverter}" Mode="OneWay" >
						
						<Binding RelativeSource="{RelativeSource AncestorType=local:DimensionButton}" Path ="Dimension" Mode="OneWay" />
						<Binding Source="{Binding}" Mode="OneWay"  />
					</MultiBinding>
				</CheckBox.IsChecked>
				<CheckBox.IsEnabled>
					<MultiBinding Converter="{StaticResource DimensionEventEnabledConverter}" Mode="OneWay">
						<Binding Source="{Binding}" Path="Content" />
						
						<Binding RelativeSource="{RelativeSource AncestorType=local:DimensionContainerCell}" Path="." />
					</MultiBinding>
				</CheckBox.IsEnabled>
			</CheckBox>
		</Border>
	</DataTemplate>
			
			
[Working WPf Repo](https://github.com/leaderanalytics/LeaderPivot.XAML.WPF)

Working WPF XAML:


    <DataTemplate x:Key="DimensionButtonItemTemplate">
        <Border Margin="2,3,2,3">
            <CheckBox HorizontalAlignment="Left" VerticalAlignment="Center"
                          Command="{Binding RelativeSource={RelativeSource AncestorType=local:DimensionButton}, Path=CheckboxCheckedCommand}"
                          CommandParameter="{Binding RelativeSource={RelativeSource TemplatedParent}, Path=Content}"
                          Content="{Binding RelativeSource={RelativeSource TemplatedParent}, Path=Content, Converter={StaticResource DimensionActionDescriptionConverter}}">

                <CheckBox.IsChecked>
                    <MultiBinding Converter="{StaticResource DimensionEventCheckedConverter}" Mode="OneWay" >
                        <Binding RelativeSource="{RelativeSource AncestorType=local:DimensionButton}" Path="Dimension" Mode="OneWay" />
                        <Binding RelativeSource="{RelativeSource TemplatedParent}" Path="Content" Mode="OneWay"  />
                    </MultiBinding>
                </CheckBox.IsChecked>
                <CheckBox.IsEnabled>
                    <MultiBinding Converter="{StaticResource DimensionEventEnabledConverter}" Mode="OneWay">
                        <Binding RelativeSource="{RelativeSource AncestorType=ListBoxItem}" Path="Content" />
                        <Binding RelativeSource="{RelativeSource AncestorType=local:DimensionContainerCell}" Path="." />
                    </MultiBinding>
                </CheckBox.IsEnabled>
            </CheckBox>
        </Border>
    </DataTemplate>

**Reply by maxkatz6** [reply]
Found that I missed your comment previously...

> <Binding Path ="{$parent[local:DimensionButton].Dimension}" ... />

Why do you add "{" and "}" inside of the binding's path?

> Command="{Binding $parent[local:DimensionButton], Path=CheckboxCheckedCommand}"

Here it also looks weird. You have duplicated "Path" property. It should be ```"{Binding $parent[local:DimensionButton].CheckboxCheckedCommand}"```. Syntax is almost the same as in WPF, except there is some Path property syntax sugar with "$parent", "$self" and "#controlname".

**Reply by sam-wheat** [reply]
@maxkatz6  Hi Max, I actually gave up on Avalonia because I could not make progress on my project.  I will open it up again and give it another try.
Why do I add curly braces...  same reason as why I am here -  I don't know what I'm doing.  These things seem obvious to you but to me not so much.

**Reply by timunie** [reply]
@sam-wheat we have tutorials and we plan to add more tutorials and samples over time. Also there is a telegram chat with community support. If that's not enough there is also paid support. 

So if you want to come back some day, you are welcome 🤗

Happy coding
Tim

**Reply by sam-wheat** [reply]
@timunie Thanks Tim I'm back at it.

**Reply by timunie** [reply]
Great, so here is the telegram chat: https://t.me/Avalonia

**Reply by sam-wheat** [reply]
I impletemented suggested fixes and still no luck.


> Furthermore, DataTemplate by itself uses TemplatedParent, which causes DimensionEventCheckedConverter to fail with "Could not bind to Dimension" exception. TemplatedParent works with Control.Template property and not ItemTemplate from what I know.

I don't know what you mean by this.

The problem I am facing is with `DimensionEventCheckedConverter`.  Specifically, the objects that are supposed to be passed to the converter via bindings are not passed and instead appear as `{(unset)}`: 

		<MultiBinding Converter="{StaticResource DimensionEventCheckedConverter}" Mode="OneWay" >
			<Binding Path="$parent[local:DimensionButton].Dimension" Mode="OneWay" />
			<Binding RelativeSource="{RelativeSource TemplatedParent}" Path="Content" Mode="OneWay" />
		</MultiBinding>
		
Working WPF XAML:

		<MultiBinding Converter="{StaticResource DimensionEventCheckedConverter}" Mode="OneWay" >
			<Binding RelativeSource="{RelativeSource AncestorType=local:DimensionButton}" Path="Dimension" Mode="OneWay" />
			<Binding RelativeSource="{RelativeSource TemplatedParent}" Path="Content" Mode="OneWay"  />
		</MultiBinding>
		
		
Links to repo and working WPF repo are above.

**Reply by timunie** [reply]
I think RelativeSource TemplatedParent is not possible in Avalonia DataTemplates. Probably you also need the parent syntax used in the binding the line above. 

I have no time to debug it on my own today.
